### DISCUSSION TITLE
Enabling distributed locking in Dynomite

### DISCUSSION THREAD

**Discussion by charybr** [Netflix/conductor#3440] [Q&A] [answered]
Hi,

We have conductor using Dynomite for persistence and queuing. Conductor uses Dyno client to connect to Dynomite. Documentation to enable redis-lock mentions value for conductor.redis-lock.serverType can be one of single, cluster or sentinel. What should be the value when using Dynomite and Dyno client.

https://github.com/Netflix/conductor/blob/main/CHANGELOG.md#redis-lock-module

Used below config:
conductor.app.workflowExecutionLockEnabled=true
conductor.workflow-execution-lock.type=redis
conductor.redis-lock.serverType=cluster
#used dynomite port
conductor.redis-lock.serverAddress=host1:8102,host2:8102,host3:8102
conductor.redis-lock.namespace=conductor

Unable to initialize locking with below error:
```
Error creating bean with name 'executionLockService' defined in URL [jar:file:/app/libs/conductor-server.jar!/BOOT-INF/lib/conductor-core-1.9.0-SNAPSHOT.jar!/com/netflix/conductor/service/ExecutionLockService.class]: Unsatisfied dependency expressed through constructor parameter 1; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'provideLock' defined in class path resource [com/netflix/conductor/redislock/config/RedisLockConfiguration.class]: Unsatisfied dependency expressed through method 'provideLock' parameter 0; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'getRedisson' defined in class path resource [com/netflix/conductor/redislock/config/RedisLockConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.redisson.Redisson]: Factory method 'getRedisson' threw exception; nested exception is java.lang.IllegalArgumentException: Redis url should start with redis:// or rediss:// (for SSL connection)
        at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:799)
        at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:228)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:1361)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFac
```

Setup:
Conductor version: 3.13.2
Persistence implementation: Dynomite
Lock: Red-Lock with Dyno client

**Comment by v1r3n** [comment, marked as answer]
Hi @charybr the locks requires either a redis server or zookeeper.  Dynomite unfortunately will not be a suitable option to be used as a locking service due to its underlying architecture/design.

**Reply by charybr** [reply]
Hi Viren,

Since we have Dynomite (which is a HA layer on top of Redis) we can use Redis for locking. We have 3-node Dynomite cluster. If we enable locking with single Redis node, will that be sufficient ?

Best Regards,
Chary

**Reply by v1r3n** [reply]
Hi @charybr In theory that should work - but this is completely untested in production so it might or not work.

**Reply by charybr** [reply]
Tried enabling the lock with Redis single node, it worked. Next we need to evaluate and then decide. Thank you so much, Viren!

**Comment by charybr** [comment]
thank you Viren! for the response. After looking at this Dyno client PR https://github.com/Netflix/dyno/pull/271, I as hoping that locking was possible with Dynomite. Adding another redis/zookeeper for locking mechanism would be on overhead. 
It would help us if you can describe those points that make Dynomite unsuitable for locking.

Best Regards,
Chary
